### ISSUE TITLE
Add granular testing capabilities

### ISSUE THREAD

**Issue by mschwager** [returntocorp/semgrep#320]
To truly feel confident in the rules I'm writing it'd be great to build on the existing testing capabilities. Excellent testing mechanisms will help prevent false positives, false negatives, and regressions. Some kind of xunit-style functionality where I can pass in code snippets and expected results and the rest is handled for me.

The filesystem tests are a great start, but it takes a while to create an exhaustive list of tests. I wonder if we could leverage something like `pytest` to pass in snippets of code from various languages to the `sgrep` runtime and grab the result. I had something like this in mind (borrowed from Dlint):

```python

BACKTRACKABLE = [
    (r"(a+)+z", True),  # present subsequent subexpression
    (r"(abc|[a-c])+z", True),  # alternation present subsequent subexpression
    (r"(a+)+", False),  # missing subsequent subexpression
    (r"(a+)+x?y?z?", False),  # all subsequent are optional
    (r"(a+)+z{0,5}", False),  # all subsequent are optional
    (r"(a+)+z{,5}", False),  # all subsequent are optional
    (r"(a+)+z*", False),  # all subsequent are optional
    (r"(a+)+b*c?d*?e", True),  # long propagation
    (r"(a+)+$", True),  # at end propagation
    (r"(a+)+\\Z", True),  # at end string propagation
    (r"(a|(bc|[b-c]))+z", True),  # parent subsequent subexpression
    (r"(a|(bc|[b-c])+z)", True),  # sibling subsequent subexpression
    (r"(a|(bc|[b-c]z)+)", False),  # sibling expression parent quantifier
    (r"(?://[^\n]*)*", False),  # found in the wild
    (r"(?:[ \t].*?(?:\n|$))*", False),  # found in the wild
    (r"<.*?>|((?:\\w[-\\w]*|&.*?;)+)", False),  # found in the wild
    (r"^(\\s+)\\w+=.*(\\n\\1\\w+=.*)+", False),  # found in the wild
    (r"([^*{}\\s]@|[^*{}@]|\\*(?!/))+", False),  # found in the wild
]
@pytest.mark.parametrize("expression, catastrophic", BACKTRACKABLE)
def test_bad_re_catastrophic_backtrackable(expression, catastrophic):
    ...
```

Basically something where we can very quickly and declaratively test out specific rules.

**Comment by aryx** [comment]
Something like that? https://github.com/returntocorp/sgrep/blob/develop/sgrep/lib/unit_matcher.ml

**Comment by mschwager** [comment]
Yes! Not sure how much work is involved to wire this up with specific rule files. In other words, some abstraction that could take a rule file, or rule id, or pattern representation, and a snippet of code and compare against an expected result.

Seems like some of the OCaml scaffolding is there, we'd just need to tie it in with the `sgrep_lint` layer above. Thoughts?

**Comment by DrewDennison** [comment]
@underyx since you're adding pytest support, figured you might like to read this issue. Feel free to unassign yourself if that makes sense

**Comment by ievans** [comment]
@mschwager is this obviated by the recent work for language graduation?

**Comment by mschwager** [comment]
> @mschwager is this obviated by the recent work for language graduation?

I don't think language graduation work solves the spirit of this task, but this task isn't very actionable without a much larger, potentially distracting, discussion/refactor so I'm fine with closing it out :+1:
